pkg/controller/policy/policy_controller.go

```diff
@@ -522,45 +557,44 @@ func (r *PolicyReconciler) completePolicy(policy *securityv1alpha1.SecurityPolic
 	return completeRules, nil
 }
 
-// getPeerGroupsAndIPBlocks get ipBlocks from groups, return unique ipBlock list
-func (r *PolicyReconciler) getPeerGroupsAndIPBlocks(namespace string, peer *securityv1alpha1.SecurityPolicyPeer) (map[string]int32, map[string]int, error) {
+// getPeersGroupsAndIPBlocks get ipBlocks from groups, return unique ipBlock list
+func (r *PolicyReconciler) getPeersGroupsAndIPBlocks(namespace string, peers []securityv1alpha1.SecurityPolicyPeer) (map[string]int32, map[string]int, error) {
 	var groups = make(map[string]int32)
 	var ipBlocks = make(map[string]int)
 
-	for group := range sets.NewString(peer.EndpointGroups...) {
-		revision, ipAddrs, exist := r.groupCache.ListGroupIPBlocks(group)
-		if !exist {
-			return nil, nil, groupNotFound(fmt.Errorf("group %s members not found", group))
-		}
-		groups[group] = revision
-
-		for _, ipBlock := range ipAddrs {
-			ipBlocks[ipBlock]++
-		}
-	}
-
-	for index := range peer.IPBlocks {
-		ipNets, err := utils.ParseIPBlock(&peer.IPBlocks[index])
-		if err != nil {
-			klog.Infof("unable parse IPBlocks %+v: %s", peer.IPBlocks, err)
-			return nil, nil, err
-		}
-		for _, ipNet := range ipNets {
-			ipBlocks[ipNet.String()]++
-		}
-	}
+	for _, peer := range peers {
+		switch {
+		case peer.Endpoint != nil:
+			var endpoint securityv1alpha1.Endpoint
+			err := r.Get(context.Background(), k8stypes.NamespacedName{Name: peer.Endpoint.Name, Namespace: peer.Endpoint.Namespace}, &endpoint)
+			if client.IgnoreNotFound(err) != nil {
+				klog.Errorf("Failed to get endpoint: %v, error: %v", peer.Endpoint, err)
+				return nil, nil, err
+			}
 
-	for _, ep := range peer.Endpoints {
-		var endpoint securityv1alpha1.Endpoint
-		ctx := context.Background()
-		err := r.Get(ctx, k8stypes.NamespacedName{Name: ep, Namespace: namespace}, &endpoint)
-		if client.IgnoreNotFound(err) != nil {
-			klog.Errorf("Failed to get endpoint: %v, error: %v", ep, err)
-			return nil, nil, err
-		}
+			for _, ip := range endpoint.Status.IPs {
+				ipBlocks[policycache.GetIPCidr(ip)]++
+			}
+		case peer.IPBlock != nil:
+			ipNets, err := utils.ParseIPBlock(peer.IPBlock)
+			if err != nil {
+				klog.Infof("unable parse IPBlock %+v: %s", peer.IPBlock, err)
+				return nil, nil, err
+			}
+			for _, ipNet := range ipNets {
+				ipBlocks[ipNet.String()]++
+			}
+		default:
+			group := peerAsEndpointGroup(namespace, peer).GetName()
+			revision, ipAddrs, exist := r.groupCache.ListGroupIPBlocks(group)
+			if !exist {
+				return nil, nil, groupNotFound(fmt.Errorf("group %s members not found", group))
```

Comment:
The default case is EndpointSelector or NamespaceSelector, it's better to add a log if not exist for specific Selector for debug purpose.

Reply:
done